inference-engine/src/mkldnn_plugin/nodes/mkldnn_reorder_node.cpp

```diff
@@ -161,14 +172,48 @@ bool MKLDNNReorderNode::created() const {
     return getType() == Reorder;
 }
 
+void MKLDNNReorderNode::optimizedNspc2Ncsp() {
+    auto parentEdge = getParentEdgeAt(0);
+    auto childEdge = getChildEdgeAt(0);
+    const int ndims = parentEdge->getDims().ndims();
+    const size_t DIM0 = parentEdge->getDims()[0];
+    const size_t DIM1 = parentEdge->getDims()[1];
+    const size_t DIM2 = ndims == 5 ? parentEdge->getDims()[ndims - 3] : 1;
+    const size_t DIM3 = parentEdge->getDims()[ndims - 2];
+    const size_t DIM4 = parentEdge->getDims()[ndims - 1];
+
+    auto src_data = reinterpret_cast<const float *>(parentEdge->getMemoryPtr()->GetPtr());
+    auto dst_data = reinterpret_cast<float *>(childEdge->getMemoryPtr()->GetPtr());
+
+    parallel_for4d(DIM0, DIM2, DIM3, DIM4, [&](size_t dim0, size_t dim2, size_t dim3, size_t dim4) {
+        for (size_t dim1 = 0; dim1 < DIM1; dim1++) {
+            size_t src_off = dim0 * DIM2 * DIM3 * DIM4 * DIM1 +
+                             dim2 * DIM3 * DIM4 * DIM1 +
+                             dim3 * DIM4 * DIM1 +
+                             dim4 * DIM1 +
+                             dim1;
+            size_t dst_off = dim0 * DIM1 * DIM2 * DIM3 * DIM4 +
+                             dim1 * DIM2 * DIM3 * DIM4 +
+                             dim2 * DIM3 * DIM4 +
+                             dim3 * DIM4 +
+                             dim4;
+            dst_data[dst_off] = src_data[src_off];
+        }
+    });
```

Comment:
```suggestion
    const size_t stride1  = DIM2 * DIM3 * DIM4;
    const size_t stride0 = stride1 * DIM1;

    parallel_for2d(DIM0, stride1, [&](size_t b, size_t j) {
        auto src_off = b*stride0 + j*DIM1;
        auto dst_off = b*stride0 + j;
        for (size_t dim1 = 0; dim1 < DIM1; ++dim1, ++src_off) {
            dst_data[dst_off] = src_data[src_off];
            dst_off += stride1;
        }
    });
```

M.b it looks less expressive, but it will slightly reduce the number of multiplications in the innermost loop.

Reply:
It won't in fact (at least gcc5.4) since it is able to optimize such patterns to just pointers increment on constant value.  Local performance experiments confirms that. However I updated the code assuming there might be compilers which are not so "smart". Thanks